### PR TITLE
Prevent crashes if freed objects are passed to SDL API functions

### PR DESCRIPTION
**Pull request by slouken**
Instead of using the magic tag in the object, we'll actually keep track of valid objects

Fixes https://github.com/libsdl-org/SDL/issues/9869
Fixes https://github.com/libsdl-org/SDL/issues/9235

### CODE REVIEW

File: src/SDL_utils.c [hunk 1]
```diff
@@ -161,6 +161,48 @@ SDL_bool SDL_ObjectValid(void *object, SDL_ObjectType type)
 void SDL_SetObjectsInvalid(void)
 {
     if (SDL_objects) {
+        SDL_assert(SDL_HashTableEmpty(SDL_objects));
```
Comment (madebr) [inline src/SDL_utils.c:164]: Perhaps move this assertion after the logging?

**Comment by madebr** [comment]
Since `SDL_Quit` and `SDL_Destroy` functions can now fail and recover, would it make sense to return an error code instead of void?

File: src/SDL_utils.c (hunk 1 above)
Comment (slouken) [reply src/SDL_utils.c]: Good idea, done!

**Comment by slouken** [comment]
> Since `SDL_Quit` and `SDL_Destroy` functions can now fail and recover, would it make sense to return an error code instead of void?

No, I don't think so.

**Comment by 1bsyl** [comment]
@slouken
I don´t think we should protect against using freed object. If that happens, it crashes and the program has to get fixed.
Anyway, the program has to get fixed and it's easier to notice a crash, than an API which silently fails.

subsidiary idea: 
if all SDL objects are put in a hashtable, then it should be possible to remove all SDL_DestroyFOO API, and keep a generic SDL_Destroy() that would work for all.

**Comment by slouken** [comment]
> @slouken I don´t think we should protect against using freed object. If that happens, it crashes and the program has to get fixed. Anyway, the program has to get fixed and it's easier to notice a crash, than an API which silently fails.

Yeah, I am planning on adding an assert (or optional assert) if the object is non-NULL and not valid.

> subsidiary idea: if all SDL objects are put in a hashtable, then it should be possible to remove all SDL_DestroyFOO API, and keep a generic SDL_Destroy() that would work for all.

Hmm, interesting idea. @icculus, thoughts?

**Comment by icculus** [comment]
I don't know, now we're keeping a hashtable to hold all objects that can't be removed for optimized builds, and a destructor function pointer for each object in that hashtable.

**Comment by slouken** [comment]
> I don't know, now we're keeping a hashtable to hold all objects that can't be removed for optimized builds, and a destructor function pointer for each object in that hashtable.

Let's go ahead and keep it as-is. We don't need to confuse things here.

**Pull request merged**